codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/OperationGenerator.java

```diff
@@ -166,18 +169,65 @@ private void constructHandler() {
      * Adds middleware to the operation middleware stack.
      */
     private void populateOperationMiddlewareStack() {
+
+        // Populate middleware's from runtime client plugins
         runtimeClientPlugins.forEach(runtimeClientPlugin -> {
             if (!runtimeClientPlugin.matchesService(model, service)
                     && !runtimeClientPlugin.matchesOperation(model, service, operation)) {
                 return;
             }
 
-            if (runtimeClientPlugin.buildMiddlewareStack().isPresent()) {
-                runtimeClientPlugin.buildMiddlewareStack().get().applyMiddleware(
-                        writer, service, operation, protocolGenerator, "stack"
-                );
+            if (runtimeClientPlugin.registerMiddleware().isPresent()) {
+                MiddlewareRegistrar middlewareRegistrar = runtimeClientPlugin.registerMiddleware().get();
+                writer.writeInline("$T(stack", middlewareRegistrar.getResolvedFunction());
+
+                Symbol functionArgument = middlewareRegistrar.getFunctionArgument();
+                if (functionArgument != null) {
+                        writer.writeInline(", $P", functionArgument);
+                }
+                writer.write(")");
             }
         });
+
+        // generate call to serde middleware helpers
+        if (protocolGenerator != null) {
+            writer.write("$L(stack)", getOperationProtocolMiddlewareHelperName(operation));
+        }
         writer.write("");
     }
+
+   /** Generate operation protocol middleware helper.
+    */
+    private void generateOperationProtocolMiddlewareHelper() {
+        if (protocolGenerator == null) {
+            return;
+        }
+        writer.openBlock("func $L (stack *middleware.Stack) {", "}",
+                getOperationProtocolMiddlewareHelperName(operation), () -> {
+
+            // Add request serializer middleware
+            String serializerMiddlewareName = ProtocolGenerator.getSerializeMiddlewareName(
+                    operation.getId(),
+                    protocolGenerator.getProtocolName());
+
+            writer.write("stack.Serialize.add(\"&$L{}\", stack)", serializerMiddlewareName);
+
+            // Adds response deserializer middleware
+            String deserializerMiddlewareName = ProtocolGenerator.getDeserializeMiddlewareName(
+                    operation.getId(),
+                    protocolGenerator.getProtocolName());
+            writer.write("stack.Deserialize.add(\"&$L{}\", stack)", deserializerMiddlewareName);
```

Comment:
I think these are missing the `middleware.After` parameter in the stack step `Add` calls.